### PR TITLE
Transpose ND HIP Kernel Optimizations

### PR DESCRIPTION
**Pull request by snehaa8**

### CODE REVIEW

File: src/modules/cpu/kernel/transpose.hpp [hunk 1]
```diff
@@ -0,0 +1,443 @@
...
+                        Rpp32u vectorLoopCount = 0;
+                        for( ; vectorLoopCount < alignedLength; vectorLoopCount += vectorIncrement)
+                        {
+                            __m256 pSrc[8];
+                            // load 64 values for source
+                            rpp_simd_load(rpp_load8_f32_to_f32_avx, srcPtrRow, &pSrc[0]);
+                            rpp_simd_load(rpp_load8_f32_to_f32_avx, srcPtrRow + 16, &pSrc[1]);
+                            rpp_simd_load(rpp_load8_f32_to_f32_avx, srcPtrRow + 32, &pSrc[2]);
+                            rpp_simd_load(rpp_load8_f32_to_f32_avx, srcPtrRow + 48, &pSrc[3]);
+                            rpp_simd_load(rpp_load8_f32_to_f32_avx, srcPtrRow + 8, &pSrc[4]);
```
Comment (r-abishek) [inline src/modules/cpu/kernel/transpose.hpp:259]: Are these actually faster than doing a generic for f32 too?

File: src/modules/cpu/kernel/transpose.hpp [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+                compute_2d_transpose(srcPtrTemp, dstPtrTemp, length[0], length[1], srcGenericDescPtr->strides[1], dstGenericDescPtr->strides[1]);
+            }
+            else if (nDim == 3)
+            {
+                // Optimized AVX version for 3D inputs of shape(x, y, 16) and permutation order (2, 0, 1)
+                if(perm[0] == 2 && perm[1] == 0 && perm[2] == 1 && length[2] == 16)
+                {
+                    Rpp32u height = length[0];
+                    Rpp32u width = length[1];
+                    Rpp32u channels = 16;
```
Comment (r-abishek) [inline src/modules/cpu/kernel/transpose.hpp:229]: Is this 16 channel version specific to a training?

**Review by r-abishek — CHANGES_REQUESTED** [review]

File: include/rppdefs.h [hunk 3]
```diff
@@ -42,7 +42,7 @@ SOFTWARE.
 /*! \brief 8 bit unsigned char maximum \ingroup group_rppdefs \page subpage_rppi */
 #define RPP_MAX_8U      ( 255 )
 /*! \brief RPP maximum dimensions in tensor \ingroup group_rppdefs \page subpage_rppt */
-#define RPPT_MAX_DIMS   ( 5 )
+#define RPPT_MAX_DIMS   ( 8 )
```
[the remark this answers is not in the document]
Comment (snehaa8) [reply include/rppdefs.h:54]: This variable is only used in RpptGenericDesc struct

File: include/rppt_tensor_geometric_augmentations.h [hunk 4]
```diff
@@ -482,6 +482,41 @@ RppStatus rppt_slice_host(RppPtr_t srcPtr, RpptGenericDescPtr srcGenericDescPtr,
...
+ * \param[in] dstGenericDescPtr destination tensor descriptor
+ * \param[in] permTensor permutation tensor for transpose operation
+ * \param[in] roiTensor ROI data for each element in source tensor (tensor of batchSize * number of dimensions * 2 values)
+ * \param [in] rppHandle RPP HOST handle created with <tt>\ref rppCreateWithBatchSize()</tt>
+ * \return A <tt> \ref RppStatus</tt> enumeration.
+ * \retval RPP_SUCCESS Successful completion.
+ * \retval RPP_ERROR* Unsuccessful completion.
+ */
+
+RppStatus rppt_transpose_generic_host(RppPtr_t srcPtr, RpptGenericDescPtr srcGenericDescPtr, RppPtr_t dstPtr, RpptGenericDescPtr dstGenericDescPtr, Rpp32u *permTensor, Rpp32u *roiTensor, rppHandle_t rppHandle);
```
[the remark this answers is not in the document]
Comment (snehaa8) [reply include/rppt_tensor_geometric_augmentations.h:599]: Modified it for HIP

File: src/modules/hip/kernel/transpose.hpp [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+{
+    int id_x = (hipBlockIdx_x * hipBlockDim_x + hipThreadIdx_x) * 8;
+    int id_y = hipBlockIdx_y * hipBlockDim_y + hipThreadIdx_y;
+
+    if(id_x >= dstStrides[0])
+        return;
+
+    int maxLength = dstStrides[0];
+    int xDiff = maxLength - (maxLength & ~7);
+    uint dstIdx = (id_y * *dstStrides++);
```
[the remark this answers is not in the document]
Comment (snehaa8) [reply src/modules/hip/kernel/transpose.hpp:22]: Sure

File: src/modules/cpu/kernel/transpose.hpp (hunk 2 above)
Comment (sampath1117) [reply src/modules/cpu/kernel/transpose.hpp:229]: Yes

File: src/modules/cpu/kernel/transpose.hpp (hunk 1 above)
Comment (sampath1117) [reply src/modules/cpu/kernel/transpose.hpp:259]: Yes

File: include/rppdefs.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (sampath1117) [reply include/rppdefs.h:54]: I feel this changed is not needed. Please revert this if not used

File: include/rppt_tensor_arithmetic_operations.h [hunk 6]
```diff
@@ -272,4 +272,4 @@ RppStatus rppt_magnitude_gpu(RppPtr_t srcPtr1, RppPtr_t srcPtr2, RpptDescPtr src
 #ifdef __cplusplus
 }
 #endif
-#endif // RPPT_TENSOR_ARITHMETIC_OPERATIONS_H
+#endif // RPPT_TENSOR_ARITHMETIC_OPERATIONS_H
```
Comment (sampath1117) [inline include/rppt_tensor_arithmetic_operations.h:275]: Please revert this change too

File: utilities/test_suite/HIP/runTests.py [hunk 7]
```diff
@@ -533,4 +533,4 @@ def rpp_test_suite_parser_and_validator():
     checkFile = os.path.isfile(qaFilePath)
     if checkFile:
         print("---------------------------------- Results of QA Test - Tensor_hip ----------------------------------\n")
-        print_qa_tests_summary(qaFilePath, supportedCaseList, nonQACaseList)
+        print_qa_tests_summary(qaFilePath, supportedCaseList, nonQACaseList)
```
Comment (sampath1117) [inline utilities/test_suite/HIP/runTests.py:536]: Please revert this change

File: utilities/test_suite/HOST/runTests.py [hunk 8]
```diff
@@ -494,4 +494,4 @@ def rpp_test_suite_parser_and_validator():
     ]
 
     for log_file in log_file_list:
-        print_performance_tests_summary(log_file, functionality_group_list, numRuns)
+        print_performance_tests_summary(log_file, functionality_group_list, numRuns)
```
Comment (sampath1117) [inline utilities/test_suite/HOST/runTests.py:497]: Please revert this change

File: utilities/test_suite/HOST/runAudioTests.py [hunk 9]
```diff
@@ -223,4 +223,3 @@ def rpp_test_suite_parser_and_validator():
     log_file_list = get_log_file_list()
     for log_file in log_file_list:
         print_performance_tests_summary(log_file, "", numRuns)
```
Comment (sampath1117) [inline utilities/test_suite/HOST/runAudioTests.py:225]: Please revert this change

**Review by sampath1117 — COMMENTED** [review]
@snehaa8 
Please address minor review comments

Also please check the latest suite for runMiscTests.py for Normalize PR and make the necessary changes in test suites

File: src/modules/hip/kernel/transpose.hpp [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+    for (int i = 0; i < dstNumDims; i++)
+    {
+        dstCoords[i].ui4[0] = (idx0123 / dstStrides[i]) % dstDims[i];
+        dstCoords[i].ui4[1] = (idx4567 / dstStrides[i]) % dstDims[i];
+    }
+    // Compute corresponding 8 srcIdxs given id_x
+    for (int i = 0; i < dstNumDims; i++)
+    {
+        for (int j = 0; j < 8; j++)
+            srcIdxs.ui1[j] += (dstCoords[permTensor[i]].ui1[j] * srcStrides[permTensor[permTensor[i]]]);
```
Comment (r-abishek) [inline src/modules/hip/kernel/transpose.hpp:42]: @snehaa8 or @sampath1117  This needs to be vectorized and the loop has to be gone:
8 uint muls -> 2 uint4 muls
8 uint adds -> 2 uint4 adds

File: src/modules/hip/kernel/transpose.hpp [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+{
+    int id_x = (hipBlockIdx_x * hipBlockDim_x + hipThreadIdx_x) * 8;
+    int id_y = hipBlockIdx_y * hipBlockDim_y + hipThreadIdx_y;
+
+    if(id_x >= dstStrides[0])
+        return;
+
+    int maxLength = dstStrides[0];
+    int xAlignedLength =  maxLength & ~7;      // alignedLength for vectorized global loads
+    int xDiff = maxLength - xAlignedLength;    // difference between roiWidth and alignedLength
```
Comment (r-abishek) [inline src/modules/hip/kernel/transpose.hpp:22]: Remove the xAlignedLength variable and just substitute with 
```
int xDiff = maxLength - (maxLength & ~7);    // difference between roiWidth and (alignedLength = maxLength & ~7)
```

File: src/modules/cpu/kernel/transpose.hpp [hunk 12]
```diff
@@ -0,0 +1,445 @@
...
+
+                                Rpp32u vectorLoopCount = 0;
+                                for( ; vectorLoopCount < alignedLength; vectorLoopCount += vectorIncrement)
+                                {
+                                    __m256 pSrc = _mm256_setr_ps(*srcPtr3,
+                                                                 *(srcPtr3 + srcGenericDescPtr->strides[1]),
+                                                                 *(srcPtr3 + 2 * srcGenericDescPtr->strides[1]),
+                                                                 *(srcPtr3 + 3 * srcGenericDescPtr->strides[1]),
+                                                                 *(srcPtr3 + 4 * srcGenericDescPtr->strides[1]),
+                                                                 *(srcPtr3 + 5 * srcGenericDescPtr->strides[1]),
```
Comment (r-abishek) [inline src/modules/cpu/kernel/transpose.hpp:363]: This whole 1/2/3/4/5/6/7 * srcGenericDescPtr->strides[1] is constant. 
Can be taken out of 4 loops!

**Comment by snehaa8** [comment]
Addressed all comments wrt to transpose HIP

File: include/rppdefs.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (sampath1117) [reply include/rppdefs.h]: This has been reverted

File: utilities/test_suite/HIP/Tensor_misc_hip.cpp [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+    fill_roi_values(nDim, batchSize, roiTensor, qaMode);
+
+    // set src/dst generic tensor descriptors
+    RpptGenericDescPtr srcDescriptorPtrND, dstDescriptorPtrND;
+    CHECK(hipHostMalloc(&srcDescriptorPtrND, sizeof(RpptGenericDesc)));
+    CHECK(hipHostMalloc(&dstDescriptorPtrND, sizeof(RpptGenericDesc)));
+    srcDescriptorPtrND->numDims = nDim + 1;
+    srcDescriptorPtrND->offsetInBytes = 0;
+    srcDescriptorPtrND->dataType = RpptDataType::F32;
+    srcDescriptorPtrND->layout = RpptLayout::NDHWC;
```
[the remark this answers is not in the document]
Comment (sampath1117) [reply utilities/test_suite/HIP/Tensor_misc_hip.cpp]: This has bean addressed in latest changes

File: utilities/test_suite/HIP/Tensor_misc_hip.cpp [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+
+    Rpp32u *permTensor;
+    if (testCase == 0)
+        CHECK(hipHostMalloc(&permTensor, nDim * sizeof(Rpp32u)));
+
+    // copy src host to device
+    CHECK(hipMemcpy(d_inputF32, inputF32, numValues * sizeof(Rpp32f), hipMemcpyHostToDevice));
+
+    // case-wise RPP API and measure time script for Unit and Performance test
+    printf("\nRunning transpose %d times (each time with a batch size of %d) and computing mean statistics...", numRuns, batchSize);
```
[the remark this answers is not in the document]
Comment (sampath1117) [reply utilities/test_suite/HIP/Tensor_misc_hip.cpp]: Addressed in latest changes

File: utilities/test_suite/HIP/Tensor_misc_hip.cpp [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+#include "../rpp_test_suite_misc.h"
+
+// fill the permuation values used for transpose
+void fill_perm_values(Rpp32u nDim, Rpp32u *permTensor, bool qaMode)
```
[the remark this answers is not in the document]
Comment (sampath1117) [reply utilities/test_suite/HIP/Tensor_misc_hip.cpp]: Addressed in latest changes

File: src/modules/cpu/kernel/transpose.hpp [hunk 16]
```diff
@@ -0,0 +1,428 @@
...
+        for(int j = 0; j < 4; j++)
+            srcPtrRow[j] = srcPtrTemp + (i + j) * srcRowStride;
+        for(int j = 0; j < 8; j++)
+            dstPtrRow[j] = dstPtrTemp + j * dstRowStride + i;
+
+        Rpp32u vectorLoopCount = 0;
+        for(; vectorLoopCount < alignedCols; vectorLoopCount += vectorIncrement)
+        {
+            __m256 pSrc[4];
+            rpp_simd_load(rpp_load8_f32_to_f32_avx, srcPtrRow[0], &pSrc[0]);
```
Comment (r-abishek) [inline src/modules/cpu/kernel/transpose.hpp:66]: @sampath1117 Pls add the AVX flags

File: src/modules/cpu/kernel/transpose.hpp [hunk 17]
```diff
@@ -0,0 +1,428 @@
...
+#include "rpp_cpu_common.hpp"
+using namespace std;
+
+void increment_ndim_ptr(Rpp32f **dstPtr, Rpp32u nDim, Rpp32u increment)
+{
+    for(int i = 0; i < nDim; i++)
+        dstPtr[i] += increment;
+}
+
+void rpp_store16_f32_f32_channelwise(Rpp32f **dstPtr, __m128 *p)
```
Comment (r-abishek) [inline src/modules/cpu/kernel/transpose.hpp:36]: Can we make these helpers inline?

**Review by r-abishek — APPROVED** [review]
